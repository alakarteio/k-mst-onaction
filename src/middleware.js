import { getPath } from 'mobx-state-tree'

const getAction = (call) => {
  const {
    name,
    type,
    context,
    args,
  } = call

  const path = getPath(context)
  const fullpath = `${path}/${name}`

  const action = {
    fullpath,
    path,
    name,
    args,
  }

  switch (type) {
    case 'process_return': return { ...action, ended: true }
    case 'action': return action
    default: return undefined
  }
}

export default dispatch => (call, next) => {
  const action = getAction(call)

  if (action) dispatch(action, call.tree)

  return next(call)
}
