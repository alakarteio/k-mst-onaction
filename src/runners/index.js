export { default as runArray } from './runArray'
export { default as runFunction } from './runFunction'
