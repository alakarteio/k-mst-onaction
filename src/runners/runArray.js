export default (dispatch, action, tree) => {
  dispatch.forEach(runner => runner(action, tree))
}
