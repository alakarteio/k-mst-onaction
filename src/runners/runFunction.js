import { isArray } from 'lodash'

export default (dispatch, action, tree) => {
  const dispatchCall = dispatch(action, tree)
  if (!dispatchCall) return
  if (isArray(dispatchCall)) dispatchCall.forEach(runner => runner(action, tree))
}
