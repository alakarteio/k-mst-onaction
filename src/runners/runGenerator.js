export default (dispatch, action, tree) => {
  const runner = dispatch(action, tree)
  let step = { done: false }

  while (!step.done) {
    step = runner.next()
    if (!step.done) step.value(action, tree)
  }
}
