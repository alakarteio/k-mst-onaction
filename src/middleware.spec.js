/* eslint-env jest */
jest.mock('./runners', () => ({
  runArray: jest.fn(),
  runFunction: jest.fn(),
}))
jest.mock('./getAction', () => call => (call.type === 'unknown' ? undefined : ({
  mocked: 'getAction',
  call,
})))

/* eslint-disable import/first */
import middleware from './middleware'
import { runArray, runFunction } from './runners'

describe('middleware', () => {
  let next
  beforeEach(() => {
    next = jest.fn(() => 'i am next')
    runArray.mockReset()
    runFunction.mockReset()
  })

  it('should not call runner (no action)', () => {
    const dispatch = jest.fn()
    const res = middleware(dispatch)({ type: 'unknown', tree: { this: 'is tree' } }, next)

    expect(res).toMatchSnapshot()
    expect(runArray.mock.calls).toMatchSnapshot()
    expect(runFunction.mock.calls).toMatchSnapshot()
  })

  it('should call runFunction', () => {
    const res = middleware(() => {})({ type: 'action', tree: { this: 'is tree' } }, next)

    expect(res).toMatchSnapshot()
    expect(runArray.mock.calls).toMatchSnapshot()
    expect(runFunction.mock.calls).toMatchSnapshot()
  })

  it('should call runArray', () => {
    const res = middleware([])({ type: 'action', tree: { this: 'is tree' } }, next)

    expect(res).toMatchSnapshot()
    expect(runArray.mock.calls).toMatchSnapshot()
    expect(runFunction.mock.calls).toMatchSnapshot()
  })

  it('should throw an error (unknown dispatch type)', () => {
    let fail = false

    try {
      middleware('dispatch string')({ type: 'action', tree: { this: 'is tree' } }, next)
    } catch (ex) {
      fail = true
      expect(ex).toMatchSnapshot()
    }

    expect(fail).toBe(true)
  })
})
