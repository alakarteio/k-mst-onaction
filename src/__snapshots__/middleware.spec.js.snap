// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware should call runArray 1`] = `"i am next"`;

exports[`middleware should call runArray 2`] = `
Array [
  Array [
    Array [],
    Object {
      "call": Object {
        "tree": Object {
          "this": "is tree",
        },
        "type": "action",
      },
      "mocked": "getAction",
    },
    Object {
      "this": "is tree",
    },
  ],
]
`;

exports[`middleware should call runArray 3`] = `Array []`;

exports[`middleware should call runFunction 1`] = `"i am next"`;

exports[`middleware should call runFunction 2`] = `Array []`;

exports[`middleware should call runFunction 3`] = `
Array [
  Array [
    [Function],
    Object {
      "call": Object {
        "tree": Object {
          "this": "is tree",
        },
        "type": "action",
      },
      "mocked": "getAction",
    },
    Object {
      "this": "is tree",
    },
  ],
]
`;

exports[`middleware should not call runner (no action) 1`] = `"i am next"`;

exports[`middleware should not call runner (no action) 2`] = `Array []`;

exports[`middleware should not call runner (no action) 3`] = `Array []`;

exports[`middleware should throw an error (unknown dispatch type) 1`] = `[Error: [trampss-mst-onaction] unknow dispatch type]`;
